lib/puppet/functions/step.rb

```diff
@@ -0,0 +1,88 @@
+# Provides stepping with given interval over elements in an iterable and optionally runs a
+# [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html) for each
+# element.
+#
+# This function takes two to three arguments.
+#
+# 1. An iterable that the function will iterate over.
+# 2. An integer step factor. This must be a positive integer.
+# 3. An optional lambda, which the function calls for each element in the interval. It must
+# request one parameter.
+#
+# @example Using the `step` function
+#
+# `$data.step(<n>) |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$stepped_data = $data.step(<n>)`
+#
+# or
+#
+# `step($data, <n>) |$parameter| { <PUPPET CODE BLOCK> }`
+#
+# or
+#
+# `$stepped_data = step($data, <n>)`
+#
+# When no third argument is present, Puppet returns an iterable that represents the nth element
+# of its first argument. This allows methods on iterables to be chained.
+#
+# When a lamdba is given as the third argument, Puppet iterates the first argument and passes each
```

Comment:
I agree that it should be called a block and that is't not an argument. In general though, I find a lot of the function documentations confusing because many of the mentions that the function takes <n> "arguments" and one of the arguments is often a "lambda" (see `each` for instance). Perhaps we should have a JIRA on how to be more consistent about this throughout in the documentation.